### PR TITLE
LtacProf (Coq v8.5)

### PR DESCRIPTION
**Pull request by JasonGross**
With the help of Paul Steckler (and @aspiwack and others at the Coq Coding Sprint), @tebbi's [LtacProf](http://www.ps.uni-saarland.de/~ttebbi/ltacprof/) now works with Coq 8.5.  The `append_end_library_hook` should possibly be done differently, but I'm not sure how it should be done.  I think this is reasonable to merge into 8.5 because it shouldn't change the behavior of any existing scripts or code, and the features it adds are used primarily for debugging, and shouldn't show up in production code (so code that works with this patch shouldn't stop working in older versions).  And having a profiler for Ltac is really, really useful.

I can squash these commits and/or remove the trailing-spaces-stripping if desired.

(Paul is currently working on porting this to trunk.)

### CODE REVIEW

**Comment by tebbi** [comment]
Hi Jason,

Thank you so much for this amazing work! I am very sorry I did not find 
the time to help.

Best,
Tobias

PS: You accidentially included the generated 
tactics/profile_ltac_tactics.ml in the commit.

**Comment by mattam82** [comment]
@aspiwack @ppedrot What do you think on the implementation side? It can't be in a pl but certainly in the upcoming release.

**Comment by ppedrot** [comment]
This looks OK to me implementationwise. What is the average overhead due to this patch when not using the debugging feature? Is is observable?

**Comment by JasonGross** [comment]
@mattam82 why shouldn't this go into a pl?  (What's the boundary between acceptable for pl vs not?  My understating was that pls can contain changes which do one of:
- fix something that is obviously a bug (in performance or behavior)
- have no effect on plugins nor developments (e.g., documentation)
- fix regressions with respect to the previous version of Coq
- add features or behavior that do not invalidate existing developments nor allow new developments which do not build with the old pl (e.g., saner error messages)
I think LtacProf basically fits into this last category, since none of the commands are meant to be used in non-testing settings.)

@ppedrot I haven't tested, but I seem to recall @tebbi mentioning a 1%-2% slowdown.  I'll plan to give you data on HoTT and the standard library later today.

**Comment by JasonGross** [comment]
The overhead on the standard library seems to be about 2.5% (about 4% on the overall build time of Coq):

Build time of successive builds of Coq (4c078b0362542908eb2fe1d63f0d867b339953fd) without LtacProf:
```
real: 167.77, user: 512.88, sys: 39.43, mem: 820740 ko
real: 165.68, user: 506.92, sys: 37.36, mem: 820624 ko
```
with LtacProf:
```
real: 172.75, user: 525.70, sys: 40.20, mem: 807528 ko
real: 171.36, user: 523.31, sys: 39.28, mem: 810588 ko
```
The command was:
```bash
make clean && git clean -xfd && ./configure -local && /usr/bin/time -f "all coq (real: %e, user: %U, sys: %S, mem: %M ko)" make STDTIME='/usr/bin/time -f "$* (real: %e, user: %U, sys: %S, mem: %M ko)"' TIMED=1 -j4 2>&1
```
The only file that changed by more than 0.6s of user time was `theories/Numbers/Rational/BigQ/QMake.v`; the changes in the slowest files (>= 5s) in the standard library are:
```
After    | File Name                                | Before   || Change
---------------------------------------------------------------------------
7m46.28s | Total                                    | 7m34.10s || +0m12.18s
---------------------------------------------------------------------------
0m45.56s | Numbers/Rational/BigQ/QMake              | 0m43.99s || +0m01.57s
0m15.96s | plugins/setoid_ring/Ncring_polynom       | 0m15.93s || +0m00.03s
0m14.29s | Numbers/Cyclic/DoubleCyclic/DoubleCyclic | 0m14.28s || +0m00.00s
0m10.57s | Numbers/Cyclic/DoubleCyclic/DoubleSqrt   | 0m10.46s || +0m00.10s
0m09.88s | FSets/FMapAVL                            | 0m09.59s || +0m00.29s
0m09.81s | MSets/MSetRBT                            | 0m09.51s || +0m00.30s
0m07.24s | Numbers/Cyclic/DoubleCyclic/DoubleDiv    | 0m07.25s || -0m00.00s
0m07.08s | Numbers/Cyclic/Int31/Cyclic31            | 0m06.97s || +0m00.11s
0m05.94s | plugins/setoid_ring/Field_theory         | 0m05.93s || +0m00.01s
0m05.75s | FSets/FMapFullAVL                        | 0m05.71s || +0m00.04s
0m05.62s | Reals/Ratan                              | 0m05.49s || +0m00.12s
0m05.58s | plugins/setoid_ring/Ring_polynom         | 0m05.46s || +0m00.12s
0m05.32s | plugins/micromega/EnvRing                | 0m05.16s || +0m00.16s
```

I suspect that most of this overhead comes from either moving more things into the tactic monad, or stringifying the names of tactics even when not profiling.  I'm currently testing a change that makes it so we only stringify tactic names when profiling, and after that I'll make a test case that stresses the calling of multiple tactics, and perhaps profile some chunk of bedrock.

**Comment by JasonGross** [comment]
The lazy stringification patch brings the overhead down to 1%, and pushes the overall compilation time down to
```
real: 169.42, user: 518.18, sys: 39.32, mem: 816676 ko
```
The log for the standard library is:
```
After    | File Name                                | Before   || Change
---------------------------------------------------------------------------
7m38.71s | Total                                    | 7m34.10s || +0m04.60s
---------------------------------------------------------------------------
0m44.59s | Numbers/Rational/BigQ/QMake              | 0m43.99s || +0m00.60s
0m16.03s | plugins/setoid_ring/Ncring_polynom       | 0m15.93s || +0m00.10s
0m14.05s | Numbers/Cyclic/DoubleCyclic/DoubleCyclic | 0m14.28s || -0m00.22s
0m10.58s | Numbers/Cyclic/DoubleCyclic/DoubleSqrt   | 0m10.46s || +0m00.11s
0m09.73s | MSets/MSetRBT                            | 0m09.51s || +0m00.22s
0m09.57s | FSets/FMapAVL                            | 0m09.59s || -0m00.01s
0m07.25s | Numbers/Cyclic/DoubleCyclic/DoubleDiv    | 0m07.25s || +0m00.00s
0m07.01s | Numbers/Cyclic/Int31/Cyclic31            | 0m06.97s || +0m00.04s
0m06.10s | plugins/setoid_ring/Field_theory         | 0m05.93s || +0m00.16s
0m05.82s | FSets/FMapFullAVL                        | 0m05.71s || +0m00.11s
0m05.59s | plugins/setoid_ring/Ring_polynom         | 0m05.46s || +0m00.12s
0m05.54s | Reals/Ratan                              | 0m05.49s || +0m00.04s
0m05.20s | plugins/micromega/EnvRing                | 0m05.16s || +0m00.04s
```

**Comment by JasonGross** [comment]
There is a bug in LtacProf that I'm not sure how to fix, though I'm not sure that this should prevent it from being merged:  LtacProf can't handle backtracking into the middle of tactics with multiple successes, because it doesn't know how to reconstruct the trace tree.  Here is a small example that gives rise to the anomaly:

```coq
Start Profiling.
Ltac multi := (idtac + idtac).
Goal True.
  try (multi; fail). (* Anomaly: Uncaught exception Failure("hd"). Please report. *)
Admitted.
```

Certainly, we could make the usage more friendly by catching this exception and printing a warning instead (what's the appropriate way to print a warning in Coq?).

I think for trunk, there should be a way to hook into the re-entering of a tactic with multiple successes, so that we can restart the timers appropriately.  ( @aspiwack @ppedrot @tebbi, thoughts on this?)

**Comment by mattam82** [comment]
@JasonGross msg_warning would be the way to print a warning

**Comment by JasonGross** [comment]
@mattam82 Thanks!  I've turned the anomaly into a warning.

**Comment by JasonGross** [comment]
What's blocking this?  @mattam82, any comments on my point in https://github.com/coq/coq/pull/150#issuecomment-209940400?

**Comment by mattam82** [comment]
I agree the feature can be considered for inclusion, but I'd like to know what @ppedrot and @aspiwack think. Couldn't you grab a few more seconds by specializing do_profile for the non-profiling case?

**Comment by ppedrot** [comment]
On a metalevel note, I am unsure that adding features after a release is a good development practice, but I am not too fond of outrageously strict rules either. Notwithstanding this point, I think we should indeed include it. What is the opinion of @maximedenes, all hail to our Great Good Development Practice Enforcer?

**Comment by gares** [comment]
@ppedrot, a good point once made by @silene is that adding a feature in plX+1 is bad if then some code working with plX+1 does not work with plX.  Eg. if we add a compilation flag in 8.5pl2, then coqc from8.5pl1 would complain (no way way to make a Makefile that works with all pl releases).

Said that, I don't know if LtacProf classifies: Start Profiling seems a debugging command to me.

**Comment by JasonGross** [comment]
I agree with the point about not adding new features in the middle of pls in general, but I think adding debugging features (and flags like -time, -profileltac, Start Profiling, etc) should be fine.

@mattam82 what do you mean by specializing do_profile?  Do you mean storing the function in a ref cell and changing which function is stored when you change the profiling behavior?

**Comment by mattam82** [comment]
No, I mean doing the condition on profiling earlier, avoiding a bind and finally

**Comment by JasonGross** [comment]
I don't think that'll work, but I don't know enough about the tactic monad to be sure; there is a "start profiling" tactic, and any tactic that comes after it needs to be profiled.  So it needs to happen at the time the tactic monad object is run, not at the thine it's assembled.  Am I misunderstanding how the monad works?

**Comment by mattam82** [comment]
Ah right, I see the bind cannot be avoided. What could be done is to do a conditional on option_start_time_add_total to avoid the finally if profiling is disabled, no?

**Comment by JasonGross** [comment]
Yes, that looks right; thanks!  I'll go do that when I get to my desktop.

**Comment by JasonGross** [comment]
@mattam82 : Done!  The overhead on building the standard library: about 0.5%

Timing stats:

Using:
```bash
make clean && git clean -xfd && ./configure -local -with-doc no -coqide no && /usr/bin/time -f "all coq (real: %e, user: %U, sys: %S, mem: %M ko)" make STDTIME='/usr/bin/time -f "$* (real: %e, user: %U, sys: %S, mem: %M ko)"' TIMED=1 -j4 2>&1
```

Without LtacProf (4c078b0), two runs:
```
real: 147.96, user: 467.93, sys: 30.80, mem: 820468 ko
real: 148.04, user: 467.92, sys: 30.49, mem: 820680 ko
```

With LtacProf, two runs:
```
real: 148.32, user: 469.09, sys: 30.57, mem: 818588 ko
real: 148.38, user: 469.71, sys: 30.56, mem: 816720 ko
```

Differences in the slowest files:
```
After    | File Name                                | Before   || Change
---------------------------------------------------------------------------
7m07.32s | Total                                    | 7m05.16s || +0m02.15s
---------------------------------------------------------------------------
0m44.90s | Numbers/Rational/BigQ/QMake              | 0m44.11s || +0m00.78s
0m15.94s | plugins/setoid_ring/Ncring_polynom       | 0m15.72s || +0m00.21s
0m14.04s | Numbers/Cyclic/DoubleCyclic/DoubleCyclic | 0m14.05s || -0m00.01s
0m10.65s | Numbers/Cyclic/DoubleCyclic/DoubleSqrt   | 0m10.45s || +0m00.20s
0m09.57s | FSets/FMapAVL                            | 0m09.53s || +0m00.04s
0m09.51s | MSets/MSetRBT                            | 0m09.45s || +0m00.06s
0m07.00s | Numbers/Cyclic/DoubleCyclic/DoubleDiv    | 0m07.05s || -0m00.04s
0m06.88s | Numbers/Cyclic/Int31/Cyclic31            | 0m06.94s || -0m00.06s
0m05.82s | plugins/setoid_ring/Field_theory         | 0m05.87s || -0m00.04s
0m05.74s | FSets/FMapFullAVL                        | 0m05.71s || +0m00.03s
0m05.37s | Reals/Ratan                              | 0m05.42s || -0m00.04s
0m05.36s | plugins/setoid_ring/Ring_polynom         | 0m05.37s || -0m00.00s
0m05.09s | plugins/micromega/EnvRing                | 0m04.96s || +0m00.12s
```

**Comment by JasonGross** [comment]
So the current status is: waiting on feedback from @maximedenes and @aspiwack ?

**Comment by maximedenes** [comment]
I think integrating this PR in 8.5 is ok with our current practices (because it is only a debugging feature). However, the danger could come from performance regressions. Testing on the stdlib doesn't seem enough, I'll try to do it on the contribs.

**Comment by herbelin** [comment]
If we take as criterion the past "pl" usages, the current request should probably more be considered as a new feature than a fix, even if related to "debugging/profiling". My understanding is that the short release policy is also going in the direction of not relaxing this usage for "pl"s.

**Comment by JasonGross** [comment]
@herbelin It would be nice to have this debugging facility in a released version of Coq that is mostly compatible with Coq 8.4.  It's currently possible to write scripts that compile in both 8.4 and 8.5 (we're doing this until we manage to port bedrock to 8.5).  If 8.6 is going to break enough things that it's going to be hard to write scripts that compile in 8.4 and also 8.6, then I'd rather see this debugging facility in 8.5.

**Comment by JasonGross** [comment]
@maximedenes if you can point me at where I can get the source for contribs that compile where 8.5pl1, I can profile them and paste the reports here.

**Comment by herbelin** [comment]
Hi Jason, I see your concern about easing the port from 8.4 to 8.5 but it would be difficult to argue that the rules are not the same as for core developers! For instance, even changes in error messages is a delicate matter (think at teaching material for instance, which might mention the text of an error and this text would change from a pl to another --- for the record, we historically adopted the plX policy so that the Coq'Art was ensured to have not only semantics but also outputs consistent when upgrading from 8.0 to 8.0plX).

So what to do for the purpose of the transition from 8.4 to 8.5? Obviously, it already works for you, so the point of releasing quickly would be that others can use it for helping porting too. Is that correct or did I miss something?

Then, how do you use it and how would you recommend to use the profiler for the purpose of the 8.4 to 8.5 transition? Do you use it to check that a new variant of a script which works both in 8.4 and 8.5 does not degrade the performance. For instance, would providing LtacProf as a plugin for 8.5 be a reasonable option or not a so good one?

Independently of LtacProf, I have questions about the 8.4 -> 8.5 transition. Do you have scripts which compile for both 8.4 and 8.5 thanks to the Coq84.v file or do you just find the right tactic invocations which works for both versions? More generally, what is your feedback at this point of the port? I spent sometimes myself working at studying how easy or not it was to port some developments and I feel it might be worth to document further the changes so as to ease the transition.

Back to the code, I'm concerned by keeping in mind some overall consistency in naming commands, a concern which in my experience is shared by many users too, simply because regular naming schemes are easier to remember than non regular ones.

For instance, the usage is more to use dashes to separate words on the command line, which should suggest a preference for naming the option -profile-ltac (or -ltac-profile).

Also, knowing that we have "Set Ltac Debug", I wondered whether this would not better be "Set Ltac Profiling" rather than "Start Profiling". However, the profiling can be suspended without being resetting, and resetting without being suspended (is that right?), so that maybe the "Set Ltac Profiling" is a bad idea here.

We shall also keep in mind that, eventually, we shall provide a "with Some Command do tactic" (see #161 ) command which would allow to write "with Start Profiling do tactic" rather than "start profiling tactic". Are you prepared to such a move eventually? (Or maybe even prepared to implement a "with" option already now?!)

In any cases, even if I did not look at the details of the implementation, relying on who already looked at the code, this looks like a very useful feature.

**Comment by JasonGross** [comment]
> I see your concern about easing the port from 8.4 to 8.5 but it would be difficult to argue that the rules are not the same as for core developers! For instance, even changes in error messages is a delicate matter (think at teaching material for instance, which might mention the text of an error and this text would change from a pl to another --- for the record, we historically adopted the plX policy so that the Coq'Art was ensured to have not only semantics but also outputs consistent when upgrading from 8.0 to 8.0plX).

Interesting.  But the argument from Coq'Art is that existing code should not behave nor print differently when upgrading.  Adding a new flag / command, which is only for debugging purposes, does not seem like it would break anything about Coq'Art, were it still being written, which could chose to either not mention the profiler, or mention that it's only available in versions > 8.*pl*.  I'm not trying to argue that LtacProf should be an exception, but that it should be part of a class of changes that are fine to merge into pls.

The transition from 8.4 to 8.5 is more annoying than I would like it to be; bedrock still only "builds" in 8.5 if I sprinkle `admit`s in many places, since I haven't found time to track down the stack overflow in the reflective automation.  This means that anything built on bedrock needs to compile in 8.4.  I also want the code to compile in 8.5, so that when we do eventually finish porting bedrock, the rest of the transition is easy.  (And getting code that compiles in 8.4 and 8.5 is not always easy, though it is usually possible.)  This is why I care about Coq 8.5 (and compatibility with Coq 8.4).
Separately, I want to be able to profile my code, and have others working on it also profile anything they build on top of it, preferably without applying custom patches to Coq.  So I want a released version of Coq that can be compatible with 8.4 that has the profiler.

> Then, how do you use it and how would you recommend to use the profiler for the purpose of the 8.4 to 8.5 transition? Do you use it to check that a new variant of a script which works both in 8.4 and 8.5 does not degrade the performance. For instance, would providing LtacProf as a plugin for 8.5 be a reasonable option or not a so good one?

It's useful for tracking down performance regressions, but it's also useful independently so long as I'm chained to 8.4 by bedrock.

So a plugin would be okay, but not ideal.  In any case, it's not currently possible to write LtacProf as a plugin, because it requires changes to tacinterp.

> Do you have scripts which compile for both 8.4 and 8.5 thanks to the Coq84.v file or do you just find the right tactic invocations which works for both versions? More generally, what is your feedback at this point of the port? I spent sometimes myself working at studying how easy or not it was to port some developments and I feel it might be worth to document further the changes so as to ease the transition.

All of Fiat compiles with both 8.4 and 8.5, thanks to `-compat 8.4` and `Compat/Coq84.v` and some makefile magic that selects which plugin source to use based on the version of Coq, and special compatibility files selected by makefile magic to fix things that haven't yet made their way into the `Coq84.v` file (such as [bug #4733](https://coq.inria.fr/bugs/show_bug.cgi?id=4733)).  I still find myself routinely writing code that breaks in 8.5 but works in 8.4 (e.g., `match constr:Set with .... end`), but overall it's not too bad to make the small changes I need to as I go along.  This is actually the biggest source of my incompatibility bug reports these days: I write code for Fiat in 8.4, it breaks in 8.5, and I go see what broke, and report a bug if it's not a deliberate change.

More generally, porting is frequently a big pain.  I have to compile the development in parallel for 8.5 and for 8.4, step through the tactics line-by-line with excruciatingly tedious copy-paste (since Ltac has no good debugger).  In the good cases, I discover that the breakage is one I already know how to fix, e.g., one of my scripts broke because `match goal with |- ?x => pose (fun x => x) end` is valid in 8.4 but not 8.5, yet the error is raised late in 8.5, leading to scripts that silently succeed but don't do as much as they used to (e.g., see [bug #4734](https://coq.inria.fr/bugs/show_bug.cgi?id=4734)).  In worse cases, I have to track down how the semantics of something changed recently (as with `Set Regular Subst Tactic`, the semantics of `;` being associative in 8.5, etc).  The absolute worst incompatibilities are the ones where things in the standard library have been renamed or moved or changed signatures between 8.4 and 8.5.  This is why I want compatibility notations   The pitfalls that come up most often, even with the `-compat 8.4` flag, are: `constr:x` works in 8.4 but not 8.5; notations in the types of hypotheses are interpreted in `type_scope` in 8.5 (so `match goal with H : context[x + y] |- _ => rewrite Plus.plus_comm in H end` is bad); names have changed in the standard library; names can no longer be shared between binders and pattern variables; and the one I dread tracking down the most, the unification algorithm has changed.

(Note that things like #153 aren't actually big deals; there's a single, obviously correct, backwards compatible change to make to my code, it's usually the case that I meant to say `proj1_sig` rather than `projT1` anyway, and I can usually catch it early).

> For instance, the usage is more to use dashes to separate words on the command line, which should suggest a preference for naming the option -profile-ltac (or -ltac-profile).

Done (https://github.com/coq/coq/pull/150/commits/b8f226f314a46906bf972dfe4297a44403756663)

> However, the profiling can be suspended without being resetting, and resetting without being suspended (is that right?), so that maybe the "Set Ltac Profiling" is a bad idea here.

Yes.  However, perhaps it would be a better idea to have `Set Ltac Profiling` and `Unset Ltac Profiling` and separately `Reset Ltac Profile` (or `Clear Ltac Profile`)?  @tebbi, what's the reason that you have `Start Profiling` also reset the profile?  I think it would be cleaner and more consistent to separate this out.  @herbelin, if you point me at the right syntax to declare such options (is it `VERNAC COMMAND EXTEND`?), and the right way to document such options in the manual (is it `\comindex`?), I'll do that unless @tebbi objects.

> Are you prepared to such a move eventually? (Or maybe even prepared to implement a "with" option already now?!)

If we make it so that `Start Profiling` (or `Set Ltac Profiling`) doesn't also reset the profile, then moving to `with Ltac Profiling do tac` (or `with Set Ltac Profiling do tac`?) will be trivial.  If you want me to declare that as the syntax of the tactic versions of these, I could do that now, if you give me the template for declaring the syntax.

**Comment by JasonGross** [comment]
A question: is there a way good way to synchronize the profile with the document state, so that backtracking reverts the profiling data?  I think this is the main reason that `Start Profiling` also resets the profile, so if backtracking worked correctly, then there'd be no need to make it do this.

**Comment by gares** [comment]
Use Summary.ref instead of simply ref to hold such a state.  Ab make sure the state is a functional data structure.

Best

**Comment by herbelin** [comment]
Thanks for your answer on porting from 8.4 to 8.5. I will comment on some specific aspects:

- Changes in library: This is also my experience that the Coq changes around NPeano requires a lot of changes in script. Is it correct that the recent 6cc5863 (May 4) improve on that?
- "match constr:Set with .... end": is the problem described somewhere?
- "constr:x": I don't understand what change you are refereeing to.
- associativity of ";": my experience is that it had a pretty limited impact on our test-suite of contributions; how big is the impact on Bedrock?
- notations in the types of hypotheses are interpreted in type_scope in 8.5: I made a comment on https://coq.inria.fr/bugs/show_bug.cgi?id=4398.
- unification: it is also my experience that changes in the unification algorithm are painful to deal with. It is not that I have so much time and being able to do something but if you have short examples illustrating the incompatibilities (additionally to the ones in the bugtracker) that would help. Are you e.g. talking about examples where the solution found is convertible but not syntactically the same as in 8.4?

**Comment by JasonGross** [comment]
@gares Thanks!  Is `type treenode = {entry : entry; children : (string, treenode) Hashtbl.t}` a functional data structure?  (I suppose not, because `Hashtbl` supports in-place modification... Is there a functional variant of `Hashtbl`?)

**Comment by gares** [comment]
You can use a Map instead of the Hashtbl.  I guess you need to use Map.Make(String) or something like that.  Are your strings long?  If not, it should be ok.

**Comment by JasonGross** [comment]
@gares The strings are probably around 100 characters at most.... they are stringified versions of tactic invocations.  Is that short enough?

**Comment by gares** [comment]
I don't know if it is a performance bottleneck, but Map uses comparison (string -> int) to give you O(log n).  So it is not about length, but about "equal prefixes" I guess.  If prefixes are similar, you may want to rev (or rev_compare) the strings.  Never optimize without measuring ;-)

**Comment by JasonGross** [comment]
@herbelin 

- https://github.com/coq/coq/commit/6cc5863b8d0e85c3d4f5e81fce7e9680134708a7 should make it easier, once I stop caring about supporting 8.5, 8.5pl1.

- `constr:x` is no longer valid in 8.5.dev? or trunk?  I can't recall where the commit is.  But by 8.6, if not already, writing parentheses in `constr:(x)` and `ltac:(x)` are required (I think this might be part of @ppedrot 's work on making Ltac less special).

- associativity of `;` bit me once or twice; I now need things like `foo; bar; bar` rather than `foo; bar` in a few places, or, `repeat repeat foo` rather than `repeat foo`.  As with most small changes in tactic semantics, it's really annoying to track down, but straightforward to fix once I pinpoint the issue.

- There are two kinds of unification problems that tend to come up: either
  1. the solution found is convertible but not syntactically the same, and so later tactic scripts that work on syntactic equality fail; or 
  2. `apply` or `rewrite` or `refine` fails to instantiate evars when there's an essentially unique, valid solution.  (I recently hit a case where `exact eq_refl` failed on a goal of the form `?n = v` for `v` an evar-free term, only if both (a) I was working in 8.5, and (b) I first ran `subst H` where `H` was in the context defined as an evar.  I can track this down if it's useful, but it's in the middle of a long and slow proof script, and failure to unify evars is one of the trickiest forms of bug to minimize, because I can't just copy-paste the goal state and start my minimizer from there.)  (I also hit another case where `apply` failed to solve a unification problem that, I believe, was trivially solvable by delaying all conversion problems and resolving evars by syntactic unification, and then checking the resulting terms for compatibility afterwards; switching to `refine` solved the problem.)

**Comment by JasonGross** [comment]
@gares can I just override the `freeze` function with one that does a deep copy of the mutable data structure?  Am I correct in inferring that this occurs between tactic invocations only?

**Comment by ppedrot** [comment]
As a general rule of thumb, I think you should go for the immutable structure instead. That removes a lot of potential interaction bugs...

**Comment by JasonGross** [comment]
@maximedenes on the Bedrock src target (with a few admits, made with -j5), the stats are:
- 8.5pl1 without ltacprof:
```
real: 1700.64, user: 2529.76, sys: 10.91, mem: 842312 ko
```
- 8.5pl1 with ltacprof
```
real: 1701.26, user: 2528.96, sys: 11.55, mem: 843320 ko
```
- 8.5pl1 with ltacprof and `-profile-ltac`:
```
real: 1737.94, user: 2576.83, sys: 17.04, mem: 839768 ko
```
Of the user time reported by the profiler, 2302.384s of it was spent in tactic-land.  This is well under 0.05% overhead when the profiler is off, and the overhead when it's turned on seems to be about 2% on tactics.

**Comment by JasonGross** [comment]
@ppedrot @gares @herbelin I've pushed a commit synchronizing the profiling data with the document state.  I spent the afternoon and evening trying to recode it as a functional data structure, and did not succeed (@tebbi's mutation-using code is too tricky for me), so I just provided a deepcopy freeze method, which seems to work in the basic tests I did.  Eventually we'll want to reimplement this as a [plugin on top of a nonlogical extension to the tactic monad that supports `Trace`, `Info`, and LtacProf](https://github.com/JasonGross/coq/tree/profiler-trace-plugin), but first I have to figure out what the plugin interface should be, and I think that's a job for 8.6 or later.

**Comment by JasonGross** [comment]
To be clear, it's now the case that `with Stop Profiling do tac`, once it exists, will do the same thing as `stop profiling; tac; start profiling`.  @herbelin, I can replace the tactics with that syntax now, if you think it's a good thing to do.

**Comment by JasonGross** [comment]
There's now `Set Ltac Profiling` and `Unset Ltac Profiling`, rather than the `Stop`/`Start`, and I've made the tactic names `start ltac profiling` and `stop ltac profiling`.

Current status of this PR, as I understand it: waiting for general feedback from @aspiwack , an ack from @gares / @ppedrot on the freeze function I wrote, comments from @maximedenes on if the profiling of bedrock src is satisfactory, and a response from @herbelin on whether I should stick with `start ltac profiling` and `stop ltac profiling` or should instead combine them into `with Set Ltac Profiling do tac` and `with Unset Ltac Profiling do tac`.

File: test-suite/success/ltacprof.v [hunk 1]
```diff
@@ -0,0 +1,8 @@
+(** Some LtacProf tests *)
+
+Set Ltac Profiling.
+Ltac multi := (idtac + idtac).
+Goal True.
+  try (multi; fail). (* Anomaly: Uncaught exception Failure("hd"). Please report. *)
```
Comment (aspiwack) [inline test-suite/success/ltacprof.v]: Is this comment up to date? I thought this had been changed to a warning.

**Comment by aspiwack** [comment]
Sorry for being late to the party.

TL;DR: I'm fine with this.

I like that it is a fairly non-intrusive change (on that subject, @JasonGross: if you use emacs, I would recommend using ws-butler (kindly available on melpa) rather than `delete-trailing-withespace` it would avoid pushing a number of irrelevant trailing whitespace changes (it is distracting for code review)).

I'm less fond of the fact that this profiler interacts purely with the Ltac layer. But as we had seen in the last coding sprint, it may not be obvious how to plug this sort of things directly within the tactic monad (@JasonGross had troubles trying to implement a tracing mechanism). So I'm ok with settling for a pragmatic solution.

I must say I am a bit concerned about the issue with backtracking (it means that if backtracking is used, even a little, in commonly used tactics, profiling may start to be mostly useless). It's probably not an issue for 8.5, but for subsequent versions of Coq, it may be useful to identify what is required for profiling to handle backtracking.

File: test-suite/success/ltacprof.v (hunk 1 above)
Comment (JasonGross) [inline test-suite/success/ltacprof.v]: I've taken to annotating test suite lines with the error messages they're supposed to prevent, to make it easier to debug their purpose if they break as Coq changes.  I'll add a "used to be: " to the front of the comment.

**Comment by JasonGross** [comment]
Thanks for the ws-butler suggestion.

Plugging the profiler into the tactic monad will eventually fix the issue with backtracking, at the cost of additional not-yet--determined overhead.  The trace info calculation is already working, all that remains there is figuring out the right interface for plugins.  But there's extra overhead there, especially when profiling (we have to compare call stacks on every call) because there's no naive way to be notified when backtracking has happened, at least none that I've seen.  So I'd rather get this merged, because it's noninvasive and low overhead and works, and update it when I next have time to work on tracing, which is rather invasive.

If we don't plug the profiler into the monad, the minimum requirement for handling backtracking is some way to be notified of the diff between the call stack before backtracking and the call stack after back tacking.  Note that the call_trace argument actually contains only a tiny segment of the call stack, and so can't be used for this purpose.

**Comment by aspiwack** [comment]
> Thanks for the ws-butler suggestion.

You are very welcome :-)

> So I'd rather get this merged, because it's noninvasive and low overhead and works

As said above, I do agree.

> If we don't plug the profiler into the monad, the minimum requirement for handling backtracking is some way to be notified of the diff between the call stack before backtracking and the call stack after back tacking. 

I'm afraid I don't really understand what you mean. Can you elaborate a bit, on how this would be used?

File: test-suite/success/ltacprof.v (hunk 1 above)
Comment (JasonGross) [inline test-suite/success/ltacprof.v]: Updated

**Comment by JasonGross** [comment]
> I'm afraid I don't really understand what you mean. Can you elaborate a bit, on how this would be used?

The profiler maintains it's own copy of the current call stack.  Whenever you enter a new tactic, you push that call on the stack; whenever you leave a tactic, you "close" the most recent stack entry, update the children of the entry above it (because this tactic you just closed is a child, and it ran for some time), and subtract off the time spent in the child from the local time that was spent in the parent itself.  The code assumes the invariant that if you wrap each tactic `tac` in `open "tac"; tclIFCATCH tac THEN close "tac" WITH exn -> (close "tac"; raise exn)`, every tactic will be opened, run, then closed.  This is not the case with reentrant (multisuccess) tactics.  The way I know to get around this is to maintain the entire call stack at all times (both in backtracking and nonbacktracking data), so that when you reenter a tactic, you can compare the stacks, and close the tactics you left, and open the tactics you reentered.

> As said above, I do agree.

Great!

Current status of this PR, as I understand it: waiting for an ack from @gares / @ppedrot on the freeze function, comments from @maximedenes on if the profiling of bedrock src is satisfactory, and a response from @herbelin on whether I should stick with `start ltac profiling` and `stop ltac profiling` or should instead combine them into `with Set Ltac Profiling do tac` and `with Unset Ltac Profiling do tac`.  (Have acks from aspiwack and mattam82.)

**Comment by JasonGross** [comment]
@gares @ppedrot @maximedenes @herbelin ?

**Comment by ppedrot** [comment]
No special comments here, I'm OK with it.

**Comment by aspiwack** [comment]
> The profiler maintains it's own copy of the current call stack. Whenever you enter a new tactic, you push that call on the stack; whenever you leave a tactic, you "close" the most recent stack entry, update the children of the entry above it (because this tactic you just closed is a child, and it ran for some time), and subtract off the time spent in the child from the local time that was spent in the parent itself. The code assumes the invariant that if you wrap each tactic tac in open "tac"; tclIFCATCH tac THEN close "tac" WITH exn -> (close "tac"; raise exn), every tactic will be opened, run, then closed. This is not the case with reentrant (multisuccess) tactics. The way I know to get around this is to maintain the entire call stack at all times (both in backtracking and nonbacktracking data), so that when you reenter a tactic, you can compare the stacks, and close the tactics you left, and open the tactics you reentered.

I see. I believe the only sane way to do that would be to keep a call stack in the monad (I believe the mechanism for Info can easily be extended for call stack).

**Comment by JasonGross** [comment]
@aspiwack are you thinking of something like [what we worked on for debug traces](https://github.com/coq/coq/compare/trunk...JasonGross:debug-command-rebased), or something else?

**Comment by JasonGross** [comment]
What's the expected timeline on this? / Who's the one to give the final green-light and merge this?

**Comment by aspiwack** [comment]
> @aspiwack are you thinking of something like what we worked on for debug traces, or something else?

Yes. Though I assume that a call stack will be more like the Info trace (which is easy) than a debug trace (which is hard).

> Who's the one to give the final green-light and merge this?

I'm guessing @ppedrot or @mattam82 are best placed for that.

**Comment by JasonGross** [comment]
Closing; it's been merged in trunk, and will not be merged in 8.5.
